compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java

```diff
@@ -0,0 +1,569 @@
+     *
+     * @param basicBlocks available basic block list of the function
+     * @param errorTableEntries available error table entries of the function
+     * @return a list of possible splits
+     */
+    private List<Split> getPossibleSplits(List<BIRBasicBlock> basicBlocks, List<BIRErrorEntry> errorTableEntries) {
+        List<Split> possibleSplits = new ArrayList<>();
+        List<BIRVariableDcl> newFuncArgs;
+        int splitEndBBIndex = basicBlocks.size() - 1; // goes from end to beginning
+        int splitEndInsIndex = basicBlocks.get(splitEndBBIndex).instructions.size() - 1;
+        int errTableEntryIndex = errorTableEntries.size() - 1;
+        boolean splitStarted = false;
+        boolean splitTypeArray = true; // will be used to mark the split caused from either NewArray or NewStructure
+        Set<BIRVariableDcl> neededOperandsVarDcl = new HashSet<>(); // that will need to be passed as function args
+        List<BIRVariableDcl> lhsOperandList = new ArrayList<>(); // that will need as localVars in the new function
+        BIROperand splitStartOperand = null;
+
+        for (int bbNum = basicBlocks.size() - 1; bbNum >= 0; bbNum--) {
+            BIRBasicBlock basicBlock = basicBlocks.get(bbNum);
+            BIRTerminator bbTerminator = basicBlock.terminator;
+            if (splitStarted) {
+                if (bbTerminator.lhsOp != null) {
+                    // as the varDcl is available inside the split no need to pass as function arg
+                    neededOperandsVarDcl.remove(bbTerminator.lhsOp.variableDcl);
+                    lhsOperandList.add(bbTerminator.lhsOp.variableDcl);
+                    if (bbTerminator.lhsOp.variableDcl.kind == VarKind.RETURN) {
+                        // if the return var is assigned value inside the split, it is not split
+                        splitStarted = false;
+                    }
+                }
+                // RHS operands varDcl will be needed to pass as function args if they won't be available inside split
+                BIROperand[] rhsOperands = bbTerminator.getRhsOperands();
+                for (BIROperand rhsOperand : rhsOperands) {
+                    if (needToPassVarDclAsArg(rhsOperand)) {
+                        neededOperandsVarDcl.add(rhsOperand.variableDcl);
+                    }
+                }
+            }
+            List<BIRNonTerminator> instructions = basicBlock.instructions;
+            for (int insNum = instructions.size() - 1; insNum >= 0; insNum--) {
+                BIRNonTerminator currIns = instructions.get(insNum);
+                if (splitStarted && (currIns.lhsOp.variableDcl.kind == VarKind.RETURN)) {
+                    // if the return var is assigned value inside the split, it is not split
+                    splitStarted = false;
+                }
+                if (splitStarted) {
+                    neededOperandsVarDcl.remove(currIns.lhsOp.variableDcl);
```

Comment:
Shall we break split started and else part to two separate methods.

Reply:
If I do that the splitStarted method would have about 13 parameters and several return values would be passed using List<Object> or a new class object. Other method also would also be like that. So may I keep as this is or may I create those 2 new methods?